Implement experimental PDF Side-by-Side feature

lyzadanger opened the pull request:
This PR is an implementation of the [requested PDF "pagefit" feature](https://github.com/hypothesis/product-backlog/issues/1046), which I'm calling "side-by-side" here for a bit more clarity.

This implementation:

* Builds off of the second sketched approach, putting all of the PDF-side-by-side logic within `PDFSidebar`
* Adds a config setting for activating this side-by-side mode (defaults to `false` at present, but I've set it explicitly to `true` in the devserver client-config defaults for easier testing)
* When configured to do so, the PDF will resize to not be overlapped by the sidebar whenever the sidebar is opened, sidebar is resized, or the window is resized
* Does not apply side-by-side mode if viewport is too small (currently 680px minimum for PDF pages, because PDF.js control layout starts breaking when narrower)
* Does not hide the sidebar on document interactions when side-by-side mode is applied
* Fixes #2372 because the code involved was right in front of me (`Guest` will no longer try to hide the sidebar before showing it when highlighted-text is clicked, eliminating a judder)
* Fixes #2463 because the code was right there...


Fixes #2477

lyzadanger commented on src/annotator/config/index.js:
```diff
@@ -22,6 +22,13 @@ export default function configFrom(window_) {
     enableExperimentalNewNoteButton: settings.hostPageSetting(
       'enableExperimentalNewNoteButton'
     ),
+    enableExperimentalPDFSideBySide: settings.hostPageSetting(
```
The naming of this config is up for all sorts of discussion. Given that we want to be more abstract about document types, there could be, e.g., an array of document types for which we should lay out side-by-side. I didn't want to overthink it for now. A single config option makes this easy to enable and test in a variety of environments.

lyzadanger commented on src/annotator/pdf-sidebar.js:
```diff
@@ -12,8 +16,101 @@ const defaultConfig = {
+      this.subscribe('sidebarLayoutChanged', state =>
+        this.fitSideBySide(state)
+      );
+      window.addEventListener('resize', () => this.fitSideBySide());
```
Do we need to debounce this?

robertknight replied:
Probably, but not essential for a first iteration.

robertknight commented on src/annotator/pdf-sidebar.js:
```diff
@@ -12,8 +16,101 @@ const defaultConfig = {
+// left in the viewport after alloting space for the sidebar is less than this
+// value, don't apply side-by-side mode, but instead allow the sidebar to
+// overlap the PDF
+const MIN_PDF_WIDTH = 320;
```
The PDF viewer toolbar has a minimum width of ~690px before the layout starts to break, so I would avoid applying side-by-side mode if the space available for the PDF viewer after applying it would be less than say 700px.

robertknight commented on src/annotator/pdf-sidebar.js:
```diff
@@ -12,8 +16,101 @@ const defaultConfig = {
+   * Override `Guest.onElementClick` to prevent the sidebar from being
+   * hidden on document clicks if in side-by-side layout
+   */
+  onElementClick(event) {
```
Rather than overriding the `onElementClick` method and then duplicating some of the logic from `Guest` to send the `hideSidebar` message, I think it would be preferable to have either a property or overridable method on the `Guest` class which determines whether auto-collapsing of the sidebar is enabled or not. If using a property, this class would update it when `this.sideBySideActive` changes. Iif using an overridable method, the `Guest` class would call it and it would return the current value of `this.sideBySideActive`.

The idea is to keep the responsibility of actually hiding the sidebar in the `Guest` class and just delegate the decision of whether or not to apply it to the subclass.

lyzadanger replied:
There were some comments in `Guest` saying that the code should really live in `Sidebar`, so I was trying to follow that design direction. I can indeed flip it around!

lyzadanger replied:
Also, just to save me a tiny bit of pain: you're not currently working on converting `Guest` to JS from coffee right now, are you?

robertknight replied:
No, I'm not working on that.

> There were some comments in Guest saying that the code should really live in Sidebar, so I was trying to follow that design direction. I can indeed flip it around!

Hmm, well perhaps it _would_ make sense for the logic to live in `Sidebar` rather than `Guest`, but more importantly it should be in just one place rather than two.

lyzadanger replied:
I agree it should be one place... I wasn't thrilled with it, as I mentioned in https://github.com/hypothesis/client/pull/2484#pullrequestreview-477000255

For now, I can put it in `Guest`. Once we've converted the full tree to JS, we should reassess, perhaps?

robertknight replied:
Yep, that sounds reasonable.

robertknight commented on src/annotator/pdf-sidebar.js:
```diff
@@ -12,8 +16,101 @@ const defaultConfig = {
+  /**
+   * Override `Guest.onElementTouchStart`
+   */
+  onElementTouchStart(event) {
```
See comments above about changing the overriding.

robertknight commented on src/annotator/pdf-sidebar.js:
```diff
@@ -12,8 +16,101 @@ const defaultConfig = {
+   * Attempt to make the PDF viewer and the sidebar fit side-by-side without
+   * overlap if there is enough room in the viewport to do so reasonably.
+   * Resize the PDF viewer container element to leave the right amount of room
+   * for the sidebar, and prompt PDFJS to re-render the PDF pages to scale
```
```suggestion
   * for the sidebar, and prompt PDF.js to re-render the PDF pages to scale
```

Per the way the project name is capitalized on the PDF.js homepage and docs.

lyzadanger replied:
Good catch.

robertknight commented on src/annotator/pdf-sidebar.js:
```diff
@@ -12,8 +16,101 @@ const defaultConfig = {
+        this.pdfContainer.style.width = maximumWidthToFit + 'px';
+      }
+      // From PDFJS `webViewerResize`
+      const currentScaleValue = this.pdfViewer.currentScaleValue;
```
I assume the purpose of this block of code is to make PDF.js adapt to the new size of `this.pdfContainer`? If so, could you add a note to that effect.

lyzadanger replied:
OK, sure. I can add the comments that PDF.js doesn't have in its own source :)

robertknight commented on src/styles/annotator/pdfjs-overrides.scss:
```diff
@@ -15,3 +15,13 @@
   background-color: rgba(0, 100, 0, 0.5);
 }
+
+// Remove horizontal `margin:auto` so that PDF pages are left-aligned
```
Hmm, I wonder whether we might want to apply this only when side-by-side mode is actually active. It looked odd to me when the sidebar is closed. The downside of conditionally applying it is that the content will "jump" more when the sidebar opens, but I think that may be better than having it off-center when trying to read rather than annotate.

lyzadanger replied:
I've reworked this in my latest round of changes.

robertknight commented on src/styles/annotator/pdfjs-overrides.scss:
```diff
@@ -15,3 +15,13 @@
+
+// Make sure sidebar bucket bar/toolbar don't obscure PDF JS tools menu
+#toolbarViewer {
+  margin-right: 32px;
```
Is there a variable that we could reference somewhere for this 32px value?

lyzadanger replied:
I doubt it: no more than there is a variable in PDF.js' CSS source, but I'll poke around just in case to see if I can be clever.

lyzadanger replied:
The magic number here is actually ours: `.annotation-frame-button` is set to `30px` (without using a SASS variable) in our own SCSS

robertknight commented on dev-server/templates/client-config.js.mustache:
```diff
@@ -2,6 +2,7 @@
   window.hypothesisConfig = function() {
     return {
       // enableExperimentalNewNoteButton: true,
+      enableExperimentalPDFSideBySide: true,
```
"Side-by-side" seems reasonable. I'm not sure that it is unambiguously better than "page fit". I don't have a better suggestion, so let's roll with it. If it proves confusing for other people we might need to reconsider. Maybe something to bikeshed on Slack.

robertknight reviewed (commented):
The overall structure looks good. I had some suggestions around the way that `Guest` and `PdfSidebar` interact and a few other details.

lyzadanger commented:
I've pushed changes that address the feedback here and a few other things I came across:

* There was a bug in the way that the sidebar width was computed when notifying on layout changes; it now includes the `Toolbar` width when relevant. This helps with the sizing of the PDF.js container
* I've reworked the CSS overrides. Now that we have a correct container width (see above), the disabling of the auto margins isn't necessary
* Regarding the specific toolbar right-margin at `30px`, that's a magic number from our own annotator SASS (which needs refactor at some point). I've commented it.
* I've reworked the click handling entirely, using a `sidebar.closeDocumentOnClick` property.
* I've changed the minimum width for side-by-side considerably based on feedback here

robertknight commented on src/annotator/pdf-sidebar.js:
```diff
@@ -12,8 +16,98 @@ const defaultConfig = {
+
+    // Prefer to lay out the sidebar and the PDF side-by-side (with the sidebar
+    // not overlapping the PDF) when space allows
+    this.sideBySide =
```
It doesn't look like this field is currently used outside of this method. Could just be a local var?

lyzadanger replied:
Ah, legacy of refactor. Sure.

lyzadanger replied:
After reflection, I've left this in place. It feels like a meaningful property on the `PdfSidebar` and makes subsequent logic more concise. It's kind of half dozen of one, six of the other...

robertknight commented on src/annotator/pdf-sidebar.js:
```diff
@@ -12,8 +16,98 @@ const defaultConfig = {
+  }
+
+  /**
+   * Set the PDF.js containiner element to the designated `width` and
```
```suggestion
   * Set the PDF.js container element to the designated `width` and
```

robertknight commented:
> I've pushed changes that address the feedback here and a few other things I came across:

These changes LGTM. I'm happy for this to land once the tests have been added/update as appropriate.

lyzadanger commented:
@LMS007 I've moved this PR from draft to "real/extreme" mode. Much of the functionality has already been greenlighted. The review needed now is of tests and also doing some more manual testing for sanity's sake.

* Tests have been added for `PDFSidebar` (where most of the changes live) and some small changes to `Sidebar` and `Guest`
* Very small code tweaks have been made; relatively negligible. I'll call them out in comments.

If it's easier to greenfield a new PR (to make this PR's history less cumbersome), let me know.

Manual testing should be (I hope) dead simple:

* Check out this branch: it sets the relevant config option to `true`, so PDF side-by-side should take effect automatically for local PDFs
* Feel free to drop additional PDF files in the `dev-server/documents/pdf` directory and they'll show up immediately (I haven't observed any PDF-document-specific behavior to this point, but it never hurts to check more).
* You're encouraged to try this out in various browsers/devices (or in browser "device mode").

LMS007 commented on src/annotator/test/pdf-sidebar-test.js:
```diff
@@ -0,0 +1,238 @@
+//import events from '../../shared/bridge-events';
+
+import PdfSidebar from '../pdf-sidebar';
+import { $imports } from '../pdf-sidebar';
```
nit these 2 lines could be 
`import PdfSidebar, { $imports } from '../pdf-sidebar';`

LMS007 replied:
edit: I think we can just remove $imports entirely.

LMS007 commented on src/annotator/test/pdf-sidebar-test.js:
```diff
@@ -0,0 +1,238 @@
+    if (!config) {
+      config = {};
+    }
+    config = Object.assign({}, sidebarConfig, config);
```
tiny nit, it would be nice to see more ES6 destructuring when we can...

```
config = {
  ...sidebarConfig, 
  ...config
}
```

lyzadanger replied:
FWIW, this was copypasta from the `sidebar-test.js` module, as the setup for testing `PDFSidebar` relies heavily on `sidebar-test` (recall that `PDFSidebar` extends `Sidebar`—there's a lot of state/setup required). But yep, I can change it.

LMS007 commented on src/annotator/test/pdf-sidebar-test.js:
```diff
@@ -0,0 +1,238 @@
+  afterEach(() => {
+    delete window.PDFViewerApplication;
+    sandbox.restore();
+    $imports.$restore();
```
Actually, I don't think we need `$imports` at all, I think it's not being mocked right?

LMS007 commented on src/annotator/test/pdf-sidebar-test.js:
```diff
@@ -0,0 +1,238 @@
+    $imports.$restore();
+  });
+
+  // sandbox.stub(window, 'innerWidth').value(1020);
```
I think you deleted this already?

LMS007 commented on src/annotator/test/pdf-sidebar-test.js:
```diff
@@ -0,0 +1,238 @@
+        window.dispatchEvent(new Event('resize'));
+
+        // PDFSidebar.fitSideBySide is invoked with no argument, so
+        // `sidebar.lastSidebarLayoutState` is used—its default values
```
"is used—its" is confusing to me, could it be re-worded?

lyzadanger replied:
Will make this clearer!

lyzadanger commented on src/annotator/test/sidebar-test.js:
```diff
@@ -506,14 +506,22 @@ describe('Sidebar', () => {
+        sinon.stub(sidebar, 'publish');
         sidebar.show();
         assert.calledOnce(layoutChangeHandlerSpy);
+        assert.calledOnce(sidebar.publish);
```
I try to avoid using an `assert.isTrue` wrapper around sinon assertions because they're harder to debug that way when they fail. If you use `isTrue` (or `isFalse`), the failure won't give you the actual failure reason (that is, it will report that it expected `true` but got `false`), whereas if I use `calledOnceWith` directly here, a test failure will output the expected and the actual args list, which is more informative.

lyzadanger commented on src/annotator/pdf-sidebar.js:
```diff
@@ -12,8 +17,103 @@ const defaultConfig = {
+   */
+  fitSideBySide(sidebarLayoutState) {
+    if (!sidebarLayoutState) {
+      sidebarLayoutState = /** @type LayoutState */ (this
```
Wowwwww, good catch, thank you

lyzadanger commented on src/annotator/test/guest-test.coffee:
```diff
@@ -346,6 +346,17 @@ describe 'Guest', ->
         assert.called(guest[methods[event]])
         assert.calledWith(guest.plugins.CrossFrame.call, 'hideSidebar')
 
+    it 'does not hide sidebar if configured not to on document click', ->
```
I will update the test description to make more sense :) but I would prefer to keep the test description about the what, not the how... (i.e. not mentioning `closeSidebarOnDocumentClick` directly). But I will try to make this more comprehensible!
